thetis/utility.py

```diff
@@ -1896,3 +1896,88 @@ def get_total_depth(self, eta):
         else:
             total_h = self.bathymetry_2d
         return total_h
+
+
+class DepthIntegratedPoissonSolver(object):
+    """
+    Construct solvers for Poisson equation and subsequently updating velocities
+
+    Generic form:
+       2D: `div(grad(q^{n+1/2})) + inner(A, grad(q^{n+1/2})) + B*q^{n+1/2} = C`
+
+    :arg A, B and C: Known functions, constants or expressions
+    :type A: vector, B: scalar, C: scalar (3D: div terms). Valued :class:`Function`, `Constant`, or an expression
+    :arg q: Non-hydrostatic pressure
+    :type q: scalar function :class:`Function`
+    """
+
+    def __init__(self, solver):
```

Comment:
It would be much better to explicitly pass in those bits of `solver` that you actually use - that way the dependencies become immediately clear. At the very least pass in the Functions that you solve for as separate arguments, so that is: `q_2d`, `uv_2d`, `elev_2d`, and `w_2d`. This code not depend on the velocity and elevation having been solved in a coupled swe system previously, so there should be no splitting here. Then `dt` and `depth` should be separate as well. `fields` then isn't necessary (batymetry is part of depth).